### PR TITLE
Segfault & memory leak hunting

### LINKED ISSUE #42: Investigate occasional segfault in test suite
I still occasionally get a segfault in the test suite so we should probably run the test suite with valgrind again to see if we can identify issues.

### PR DESCRIPTION
**Pull request by Cadair**
fixes #42

### CODE REVIEW

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 1]
```diff
@@ -256,20 +262,34 @@ static PyObject *compress_hcompress_1_c(PyObject *self, PyObject *args) {
...
+
   // maxelem adapted from cfitsio's imcomp_calc_max_elem function
   maxelem = count / 4 * 2.2 + 26;
 
   compressed_values = (char *)malloc(maxelem);
 
   if (bytepix == 4) {
     decompressed_values_int = (int *)str;
-    fits_hcompress(decompressed_values_int, ny, nx, scale, compressed_values, &count, &status);
+    fits_hcompress(decompressed_values_int, ny, nx, scale, compressed_values, &maxelem, &status);
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:278]: I am not 100% sure this is correct.

The comment in `fits_hcompress` says:
>    nbyts  - input value = size of the output buffer;
>            returned value = size of the compressed byte stream, in bytes

which makes me think it should be this?

**Comment by astrofrog** [comment]
I investigated a bit and ran into the error:

```
E       ValueError: encode: output buffer too small
```

for:

```
 test_compress[HCOMPRESS_1 {} <i2-shape: (12, 12) tile_dims: (4, 5)]
```

I dug a bit deeper and found that in the qwrite function in CFITSIO the comparison:

```
     if (noutchar + n > noutmax) {
```

is behaving weirdly - for example for: noutchar=0, n=3 and noutmax=110, noutchar  + n > noutmax evaluates as true even though it should be false. However, casting both sides in the inequality to (long) somehow fixes it:

```
     if ((long)(noutchar + n) > (long)noutmax) {
		return(0);  /* buffer overflow */
	 }
```

**Comment by astrofrog** [comment]
Ok I figured that one out and have pushed a commit, there was a simple mistyping in our C wrappers.

**Comment by astrofrog** [comment]
One of the remaining errors is:

```
        if compression_type == "HCOMPRESS_1":
            if self._image_header["NAXIS1"] < 4 or self._image_header["NAXIS2"] < 4:
                raise ValueError("Hcompress minimum image dimension is 4 pixels")
            elif tile_size:
                if tile_size[0] < 4 or tile_size[1] < 4:
                    # user specified tile size is too small
>                   raise ValueError("Hcompress minimum tile dimension is 4 p
```

is that just that we need to make this check not use [0] and [1] but actually check the non-1 dimensions of the tile_size? (i.e. for a tile shape of [1, 5, 4] we should check that exactly two values are >=4 and the rest are 1? I pushed a commit to fix this, but feel free to revert if we shouldn't.

**Comment by astrofrog** [comment]
The last error I see is:

```
E       ValueError: HCOMPRESS requires tiles of at least 4x4 pixels.
```

which occurs with an edge tile in HCOMPRESS (where the edge tile is <4x4). I'm not sure how we deal with that - if that is really true do we need to pad the data? or is the check incorrect?

**Comment by Cadair** [comment]
> is that just that we need to make this check not use [0] and [1] but actually check the non-1 dimensions of the tile_size? 

Yes I think so. That's what we have where we calculate the nx ny arguments to the codec.

**Comment by astrofrog** [comment]
This code in hcompress puzzles me a bit:

```
	    
            if (zbitpix == BYTE_IMG || zbitpix == SHORT_IMG) {
                fits_hcompress(idata, tilenx, tileny, 
		  ihcompscale, (char *) cbuf, &hcomp_len, status);

            } else {
                 /* have to convert idata to an I*8 array, in place */
                 /* idata must have been allocated large enough to do this */

                fits_int_to_longlong_inplace(idata, tilelen, status);
                lldata = (LONGLONG *) idata;		

                fits_hcompress64(lldata, tilenx, tileny, 
		  ihcompscale, (char *) cbuf, &hcomp_len, status);
            }
```

because it looks like ``fits_hcompress`` is only used for byte or short, and for any other types fits_hcompress64 is used

**Comment by astrofrog** [comment]
Ok so it turns out the C code really can't real with tile sizes <4x4 - but I realized that in fact ``CompImageHDU`` checks to make sure that no edge tiles are <4x4 and will raise an error ("Last tile along 1st dimension has less than 4 pixels") - it's just that that code was written to check the first two dimensions and needed to be generalized.

See https://github.com/aperiosoftware/astropy/pull/49/commits/e0533e5c9adf657f1d75317381d0b441c9fd4451 for a generalized version of this which seems to get rid of the errors complaining about too small tiles.

This brings the number of failed/errored tests down from 23 to 10/11

**Comment by astrofrog** [comment]
A lot of the remaining failures are related to i4 dtype and some mention overflow. I wonder whether as mentioned in https://github.com/aperiosoftware/astropy/pull/49#issuecomment-1355715258 we should be using the '64' function for i4? (not clear).

**Comment by astrofrog** [comment]
As discussed on Matrix I think we probably should back out the changes here to allow arbitrary tile order for HCOMPRESS, and also only test tile shapes for HCOMPRESS where the data is a multiple of the number of tiles. I've opened #51 to remind us to add support for this afterwards.

**Comment by Cadair** [comment]
@astrofrog I think we are good to go here and this is :crossed_fingers:  the last of the segfaults.

**Review by astrofrog — APPROVED** [review]

**Pull request merged**